ape_etherscan/explorer.py

```diff
@@ -1,14 +1,46 @@
+from json.decoder import JSONDecodeError
 from ape.api import ExplorerAPI
+from ape.types import ABI, ContractType
+from typing import Optional
+import requests
+import json
 
 ETHERSCAN_URI = (
     lambda n: (f"https://{n}.etherscan.io/" if n != "mainnet" else "https://etherscan.io/")
     + "{0}/{1}/"
 )
 
 
+def get_etherscan_uri(network_name):
+    return (
+        f"https://api-{network_name}.etherscan.io/api"
+        if network_name != "mainnet"
+        else "https://api.etherscan.io/api"
+    )
+
+
 class Etherscan(ExplorerAPI):
     def get_address_url(self, address: str) -> str:
         return ETHERSCAN_URI(self.network.name).format("address", address)
 
     def get_transaction_url(self, transaction_hash: str) -> str:
         return ETHERSCAN_URI(self.network.name).format("tx", transaction_hash)
+
+    def get_contract_type(self, address: str) -> Optional[ContractType]:
+        response = requests.get(
+            get_etherscan_uri(self.network.name),
+            params={"module": "contract", "action": "getsourcecode", "address": address},
+        )
+        result = response.json().get("result")
+        if not result or len(result) != 1:
```

Comment:
When would the length be greater than 1?

Reply:
Shouldn't be, but just in case, doesn't hurt to add the check